pkg/virtctl/update/machine-type/convert-machine-type/controller.go

```diff
@@ -0,0 +1,260 @@
+		fmt.Println("Timed out waiting for caches to sync")
+		return
+	}
+
+	vmKeys := c.VmInformer.GetStore().ListKeys()
+	for _, k := range vmKeys {
+		c.Queue.Add(k)
+	}
+
+	wait.Until(c.runWorker, time.Second, stopCh)
+}
+
+func (c *JobController) runWorker() {
+	for c.Execute() {
+	}
+}
+
+func (c *JobController) Execute() bool {
+	key, quit := c.Queue.Get()
+	if quit {
+		close(c.ExitJob)
+		return false
+	}
+
+	defer c.Queue.Done(key)
+
+	if err := c.execute(key.(string)); err != nil {
+		c.Queue.AddRateLimited(key)
+	} else {
+		c.Queue.Forget(key)
+		c.exitJob()
+	}
+
+	return true
+}
+
+func (c *JobController) vmHandler(obj interface{}) {
+	key, err := cache.DeletionHandlingMetaNamespaceKeyFunc(obj)
+	if err == nil {
+		c.Queue.Add(key)
+	}
+}
+
+func (c *JobController) execute(key string) error {
+	obj, exists, err := c.VmInformer.GetStore().GetByKey(key)
+	if err != nil {
+		return err
+	}
+
+	if !exists {
+		return fmt.Errorf("VM does not exist")
+	}
+
+	vm := obj.(*v1.VirtualMachine)
+
+	// check if VM is running
+	isRunning, err := vmIsRunning(vm)
+	if err != nil {
+		fmt.Println(err)
+		return err
+	}
+
+	// check if VM machine type was updated
+	updated, err := isMachineTypeUpdated(vm)
+	if err != nil {
+		fmt.Println(err)
+		return err
+	}
+
+	// update VMs that require update
+	if !updated {
+		err := c.patchMachineType(vm)
+		if err != nil {
+			return err
+		}
+
+		// don't need to do anything else to stopped VMs
+		if !isRunning {
+			return nil
+		}
+
+		// if force restart flag is set, restart running VMs immediately
+		// don't apply warning label to VMs being restarted
+		if RestartNow {
+			return c.VirtClient.VirtualMachine(vm.Namespace).Restart(context.Background(), vm.Name, &v1.RestartOptions{})
+		}
+
+		// adding the warning label to the running VMs to indicate to the user
+		// they must manually be restarted
+		patchString := `[{ "op": "add", "path": "/status/machineTypeRestartRequired", "value": true }]`
+		err = c.statusUpdater.PatchStatus(vm, types.JSONPatchType, []byte(patchString), &metav1.PatchOptions{})
+		if err != nil {
+			return err
+		}
+	}
+
+	if isRunning {
+		// check if VMI machine type has been updated
+		updated, err = c.vmiIsUpdated(vm)
+		if err != nil {
+			fmt.Println(err)
+			return err
+		}
+
+		if !updated {
+			fmt.Println("vmi machine type has not been updated")
+			return nil
+		}
+	}
+
+	// mark MachineTypeRestartRequired as false
+	patchString := `[{ "op": "replace", "path": "/status/machineTypeRestartRequired", "value": false }]`
+	err = c.statusUpdater.PatchStatus(vm, types.JSONPatchType, []byte(patchString), &k8sv1.PatchOptions{})
+	if err != nil {
+		fmt.Println(err)
+		return err
+	}
+
+	return nil
+}
+
+func vmIsRunning(vm *v1.VirtualMachine) (bool, error) {
+	runStrategy, err := vm.RunStrategy()
+	if err != nil {
+		return false, err
+	}
+
+	if runStrategy == v1.RunStrategyAlways {
+		return true, nil
+	}
+
+	return false, nil
+}
```

Comment:
I am not sure if it is better to check the status condition instead of the spec.